src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractAddItemCommandHandler.cs

```diff
@@ -0,0 +1,122 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.Input;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands
+{
+    /// <summary>
+    /// Provides support for all Add Item commands that operate on <see cref="IProjectTree"/> nodes, across C# and VB
+    /// </summary>
+    internal abstract partial class AbstractAddItemCommandHandler : IAsyncCommandGroupHandler
+    {
+        protected static readonly Guid LegacyCSharpPackageGuid = new Guid("{FAE04EC1-301F-11d3-BF4B-00C04F79EFBC}");
+        protected static readonly Guid LegacyVBPackageGuid = new Guid("{164B10B9-B200-11d0-8C61-00A0C91E29D5}");
+        private readonly ConfiguredProject _configuredProject;
+        private readonly IPhysicalProjectTree _projectTree;
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IVsUIService<IVsAddProjectItemDlg> _addItemDialog;
+        private readonly IVsUIService<IVsShell> _vsShell;
+        private readonly Lazy<Dictionary<long, List<TemplateDetails>>> _commandMap;
+
+        [ImportingConstructor]
+        public AbstractAddItemCommandHandler(ConfiguredProject configuredProject, IPhysicalProjectTree projectTree, IUnconfiguredProjectVsServices projectVsServices, IVsUIService<IVsAddProjectItemDlg> addItemDialog, IVsUIService<SVsShell, IVsShell> vsShell)
+        {
+            _configuredProject = configuredProject;
+            _projectTree = projectTree;
+            _projectVsServices = projectVsServices;
+            _addItemDialog = addItemDialog;
+            _vsShell = vsShell;
+
+            _commandMap = new Lazy<Dictionary<long, List<TemplateDetails>>>(() => GetTemplateDetails());
+        }
+
+        protected abstract Dictionary<long, List<TemplateDetails>> GetTemplateDetails();
+
+        public Task<CommandStatusResult> GetCommandStatusAsync(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, string commandText, CommandStatus progressiveStatus)
+        {
+            Requires.NotNull(nodes, nameof(nodes));
+
+            if (nodes.Count == 1 && _projectTree.NodeCanHaveAdditions(nodes.First()) && TryGetCommandDetails(commandId, out _))
+            {
+                return GetCommandStatusResult.Handled(commandText, CommandStatus.Enabled);
+            }
+
+            return GetCommandStatusResult.Unhandled;
+        }
+
+        public async Task<bool> TryHandleCommandAsync(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut)
+        {
+            Requires.NotNull(nodes, nameof(nodes));
+
+            if (nodes.Count == 1)
+            {
+                if (TryGetCommandDetails(commandId, out TemplateDetails result))
```

Comment:
Can you merge the two if blocks or move to early exit to avoid unneeded nesting.